File: {{cookiecutter.project_slug}}/backend-node/src/routes/todos.ts
import { Router } from 'express'
import type { Todo } from '../types/index.js'

const router = Router()

// In-memory storage for demo purposes
let todos: Todo[] = []
let nextId = 1

// GET /api/v1/todos
router.get('/', (_req, res) => {
  res.json({ data: todos, error: null })
})

// GET /api/v1/todos/:id
router.get('/:id', (req, res) => {
  const todo = todos.find((t) => t.id === parseInt(req.params.id))
  if (!todo) {
    res.status(404).json({ data: null, error: 'Todo not found' })
    return
  }
  res.json({ data: todo, error: null })
})

// POST /api/v1/todos
router.post('/', (req, res) => {
  const { text } = req.body
  if (!text || typeof text !== 'string') {
    res.status(400).json({ data: null, error: 'Text is required' })
    return
  }

  const todo: Todo = {
    id: nextId++,
    text: text.trim(),
    completed: false,
  }
  todos.push(todo)
  res.status(201).json({ data: todo, error: null })
})

// PATCH /api/v1/todos/:id
router.patch('/:id', (req, res) => {
  const id = parseInt(req.params.id)
  const todoIndex = todos.findIndex((t) => t.id === id)

  if (todoIndex === -1) {
    res.status(404).json({ data: null, error: 'Todo not found' })
    return
  }

  const { text, completed } = req.body
  if (text !== undefined) {
    todos[todoIndex].text = text
  }
  if (completed !== undefined) {
    todos[todoIndex].completed = completed
  }

  res.json({ data: todos[todoIndex], error: null })
})

// DELETE /api/v1/todos/:id
router.delete('/:id', (req, res) => {
  const id = parseInt(req.params.id)
  const todoIndex = todos.findIndex((t) => t.id === id)

  if (todoIndex === -1) {
    res.status(404).json({ data: null, error: 'Todo not found' })
    return
  }

  const [deleted] = todos.splice(todoIndex, 1)
  res.json({ data: deleted, error: null })
})

export { router as todoRouter }

File: {{cookiecutter.project_slug}}/backend-python/app/routes/todos.py
from typing import Any

from flask import Blueprint, jsonify, request

bp = Blueprint("todos", __name__)

# In-memory storage for demo purposes
todos: list[dict[str, Any]] = []
next_id = 1


@bp.get("/")
def get_todos() -> tuple[dict[str, Any], int]:
    """Get all todos."""
    return {"data": todos, "error": None}, 200


@bp.get("/<int:todo_id>")
def get_todo(todo_id: int) -> tuple[dict[str, Any], int]:
    """Get a specific todo by ID."""
    todo = next((t for t in todos if t["id"] == todo_id), None)
    if not todo:
        return {"data": None, "error": "Todo not found"}, 404
    return {"data": todo, "error": None}, 200


@bp.post("/")
def create_todo() -> tuple[dict[str, Any], int]:
    """Create a new todo."""
    global next_id

    data = request.get_json()
    if not data or not data.get("text"):
        return {"data": None, "error": "Text is required"}, 400

    todo = {
        "id": next_id,
        "text": data["text"].strip(),
        "completed": False,
    }
    next_id += 1
    todos.append(todo)

    return {"data": todo, "error": None}, 201


@bp.patch("/<int:todo_id>")
def update_todo(todo_id: int) -> tuple[dict[str, Any], int]:
    """Update a todo."""
    todo = next((t for t in todos if t["id"] == todo_id), None)
    if not todo:
        return {"data": None, "error": "Todo not found"}, 404

    data = request.get_json()
    if data:
        if "text" in data:
            todo["text"] = data["text"]
        if "completed" in data:
            todo["completed"] = data["completed"]

    return {"data": todo, "error": None}, 200


@bp.delete("/<int:todo_id>")
def delete_todo(todo_id: int) -> tuple[dict[str, Any], int]:
    """Delete a todo."""
    global todos

    todo_index = next((i for i, t in enumerate(todos) if t["id"] == todo_id), None)
    if todo_index is None:
        return {"data": None, "error": "Todo not found"}, 404

    deleted = todos.pop(todo_index)
    return {"data": deleted, "error": None}, 200
